In [1]:
import os
import pandas as pd
import numpy as np
import pandas_profiling
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LassoCV
from sklearn import datasets
from collections import Counter

from joblib import dump
from sklearn.externals import joblib

#from keras.models import model_from_json

#Import Gaussian Naive Bayes model
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB

#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import cross_val_score


C:\Users\KaiLo\Anaconda3\lib\site-packages\sklearn\externals\joblib\__init__.py:15: DeprecationWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=DeprecationWarning)


# Se importa el Dataset con informacion de nubes
Como se ha descrito, las imagenes descargadas de los portales de imagenes satelitales presentan información en multiples bandas. Es con esta información que los algoritmos de clasificación pueden emitir un juicio para determinar a que tipo de información puede corresponder un pixel: bosque, no busque,nubes,agua, agua_bosque,ninguno.
En este escenario se plantea probar el clasificador Naive Bayes con información de imagenes que contengan nubes y agregar una nueva clase: Nubes. 
Este escenario se plantea en respuesta a que la clasificación de imagenes satelitales cuyas nubes fueron removidas resultaban un problema facil de clasificar para este tipo de modelo. 

In [2]:
path = './satellite_dataset-con-nubes.csv'

satellite = pd.read_csv(path)
satellite.head(100)

blue    green      red      nir   swir1   swir2  wofs  bosque  \
0     286.0    197.0     96.0     83.0   116.0    95.0     1       0   
1     280.0    217.0    108.0     83.0   118.0    96.0     1       0   
2    3011.0   2996.0   2900.0   4186.0  1683.0  1373.0     0       0   
3    3576.0   3303.0   3313.0   3369.0  2637.0  2339.0     0       0   
4    1398.0   1345.0   1311.0   1389.0  1633.0  1355.0     0       0   
5     328.0    232.0    118.0     96.0   124.0    95.0     1       0   
6     230.0    203.0    105.0     96.0   129.0   105.0     1       0   
7     218.0    317.0    142.0    125.0   169.0   150.0     1       0   
8    6702.0   6782.0   7018.0   7304.0  6323.0  5154.0     0       0   
9    2866.0   2807.0   2806.0   2901.0  3267.0  2668.0     0       0   
10   4682.0   4869.0   4784.0   5231.0  4318.0  3072.0     0       0   
11    241.0    266.0    127.0   1156.0   341.0   129.0     0       1   
12    100.0    254.0    142.0   1004.0   456.0   236.0     0       1   
13   5969.0   6137.0   6251.0   7166.0  5648.0  4379.0     0       0   
14   6500.0   6588.0   6696.0   6850.0  5156.0  4107.0     0       0   
15   1541.0   1450.0   1409.0   2846.0  1827.0  1268.0     0       0   
16   1826.0   1623.0   1593.0   1635.0  1432.0  1197.0     0       0   
17    178.0    465.0    264.0   4109.0  1557.0   534.0     0       1   
18   4673.0   4564.0   4520.0   5229.0  3554.0  2615.0     0       0   
19   1573.0   1560.0   1508.0   1522.0  1665.0  1484.0     0       0   
20   2756.0   2676.0   2691.0   5450.0  3509.0  2345.0     0       0   
21    225.0    207.0    111.0    102.0   143.0   110.0     1       0   
22    374.0    400.0    260.0    229.0   241.0   191.0     1       0   
23    950.0    734.0    608.0    575.0   523.0   424.0     1       0   
24    398.0    317.0    189.0    147.0   154.0   115.0     1       0   
25    150.0    317.0    179.0   2665.0  1078.0   398.0     0       1   
26   9659.0   9833.0  10162.0  10176.0  4104.0  2899.0     0       0   
27    192.0    109.0     16.0      5.0    47.0    42.0     1       0   
28    176.0     96.0      8.0      1.0    49.0    45.0     1       0   
29   2986.0   3055.0   3005.0   4823.0  3440.0  2584.0     0       0   
..      ...      ...      ...      ...     ...     ...   ...     ...   
70    111.0    376.0    167.0   2833.0  1009.0   330.0     0       1   
71    513.0    412.0    307.0    279.0   323.0   243.0     1       0   
72   3418.0   3443.0   3520.0   3703.0  4167.0  3728.0     0       0   
73   2350.0   2066.0   2090.0   2171.0  2034.0  1960.0     0       0   
74    514.0    605.0    498.0   1874.0   940.0   576.0     0       0   
75   1916.0   1749.0   1704.0   1732.0  1727.0  1617.0     0       0   
76    243.0    161.0     58.0     45.0    76.0    61.0     1       0   
77    284.0    232.0    128.0    111.0   137.0   112.0     1       0   
78    261.0    215.0    123.0     90.0   116.0    93.0     1       0   
79    267.0    189.0     73.0     46.0    75.0    61.0     1       0   
80    293.0    226.0    124.0    105.0   128.0   101.0     1       0   
81    877.0    926.0    754.0    705.0   860.0   777.0     0       0   
82   1845.0   1748.0   1678.0   1661.0  1715.0  1630.0     0       0   
83    215.0    146.0     45.0     18.0    58.0    44.0     1       0   
84    802.0   1097.0    931.0   2167.0  1557.0  1021.0     0       0   
85   1051.0    814.0    731.0    705.0   563.0   473.0     0       0   
86   -354.0    328.0    217.0    726.0  1185.0   920.0     0       0   
87   3984.0   3828.0   3914.0   4109.0  4272.0  3917.0     0       0   
88   8936.0   8796.0   8885.0   8324.0  3226.0  2161.0     0       0   
89   3823.0   3739.0   3804.0   3966.0  4244.0  3657.0     0       0   
90     69.0    186.0     89.0   1299.0   465.0   158.0     0       1   
91   7384.0   7729.0   7911.0   8104.0  6644.0  5029.0     0       0   
92  10840.0  11634.0  11979.0  12126.0  5792.0  3701.0     0       0   
93    149.0     84.0     -7.0    -19.0    35.0 

# Filtro de valores negativos
Se realiza una filtro de las columnas que en alguna banda contengan un valor negativo porque el clasificador multinomial Naive bayes no funciona bien con esta información.

In [3]:
satellite = satellite[satellite["blue"] > 0]
satellite = satellite[satellite["red"] > 0]
satellite = satellite[satellite["green"] > 0]
satellite = satellite[satellite["nir"] > 0]
satellite = satellite[satellite["swir1"] > 0]
satellite = satellite[satellite["swir2"] > 0]

In [4]:
#satellite['wofs'].value_counts()
#satellite['bosque'].value_counts()
#satellite['wofs_bosque'].value_counts()
#satellite['cloud'].value_counts()
#satellite['ninguno'].value_counts()

# Se reemplaza los valores en columnas de clases
Como parte de preparación de los datos para crear el modelo se deben hacer algunos ajustes a la forma en que el dataset muestra la infiromación.Como se puede observar el dataset, las columnas de las clases se encuentran categorizadas por un valor binario. Lo que se realiza a continuación es reemplazar estos valores binarios por NaN en caso de ser "Cero" y el nombre de la clase en caso de ser "Uno".

In [5]:
# Reemplazar el valor de columna Wofs por NaN(0) y Wofs(1)
satellite['wofs'] = satellite['wofs'].replace(0,np.nan)
satellite['wofs'] = satellite['wofs'].replace(1,'wofs')


In [6]:
# Reemplazar el valor de columna bosque por NaN(0) y bosque(1)
satellite['bosque'] = satellite['bosque'].replace(0,np.nan)
satellite['bosque'] = satellite['bosque'].replace(1,'bosque')

In [7]:
# Reemplazar el valor de columna wofs_bosque por NaN(0) y wofs_bosque(1)
satellite['wofs_bosque'] = satellite['wofs_bosque'].replace(0,np.nan)
satellite['wofs_bosque'] = satellite['wofs_bosque'].replace(1,'wofs_bosque')

In [8]:
# Reemplazar el valor de columna ninguno por NaN(0) y ninguno(1)
satellite['ninguno'] = satellite['ninguno'].replace(0,np.nan)
satellite['ninguno'] = satellite['ninguno'].replace(1,'ninguno')

In [9]:
# Reemplazar el valor de columna ninguno por NaN(0) y cloud(1)
satellite['cloud'] = satellite['cloud'].replace(0,np.nan)
satellite['cloud'] = satellite['cloud'].replace(1,'cloud')

# Nueva columna de clasificación
Como ultima parte para la preparación de los datos debemos tener la clasificación en una sola columna. Procedemos a crearla con el nombre "label" y alli asignaremos el valor categorizado de cada conjunto de atributos.

In [10]:
#Se crea una nueva columna para almacenar el valor de etiqueta
satellite['label']=np.nan

#Se indica en la columna 'label', la clasificacion de la columna 'wofs'. Se esribe "Wofs" en columna 'label' si en  
#la columna 'wofs' esta este valor
for label in [col for col in satellite.columns if 'wofs' in col]:
    satellite['label'].fillna(satellite[label],inplace=True)
    
#Se indica en la columna 'label', la clasificacion de la columna 'bosque'. Se esribe "bosque" en columna 'label' si en  
#la columna 'bosque' esta este valor
for label in [col for col in satellite.columns if 'bosque' in col]:
    satellite['label'].fillna(satellite[label],inplace=True)
    
#Se indica en la columna 'label', la clasificacion de la columna 'wofs_bosque'. Se esribe "wofs_bosque" en columna 'label' si en  
#la columna 'wofs_bosque' esta este valor
for label in [col for col in satellite.columns if 'wofs_bosque' in col]:
    satellite['label'].fillna(satellite[label],inplace=True)
    
#Se indica en la columna 'label', la clasificacion de la columna 'ninguno'. Se esribe "ninguno" en columna 'label' si en  
#la columna 'ninguno' esta este valor
for label in [col for col in satellite.columns if 'ninguno' in col]:
    satellite['label'].fillna(satellite[label],inplace=True)
    
#Se indica en la columna 'label', la clasificacion de la columna 'cloud'. Se esribe "cloud" en columna 'label' si en  
#la columna 'cloud' esta este valor
for label in [col for col in satellite.columns if 'cloud' in col]:
    satellite['label'].fillna(satellite[label],inplace=True)
satellite.head(20)

blue   green     red     nir   swir1   swir2  wofs  bosque wofs_bosque  \
0    286.0   197.0    96.0    83.0   116.0    95.0  wofs     NaN         NaN   
1    280.0   217.0   108.0    83.0   118.0    96.0  wofs     NaN         NaN   
2   3011.0  2996.0  2900.0  4186.0  1683.0  1373.0   NaN     NaN         NaN   
3   3576.0  3303.0  3313.0  3369.0  2637.0  2339.0   NaN     NaN         NaN   
4   1398.0  1345.0  1311.0  1389.0  1633.0  1355.0   NaN     NaN         NaN   
5    328.0   232.0   118.0    96.0   124.0    95.0  wofs     NaN         NaN   
6    230.0   203.0   105.0    96.0   129.0   105.0  wofs     NaN         NaN   
7    218.0   317.0   142.0   125.0   169.0   150.0  wofs     NaN         NaN   
8   6702.0  6782.0  7018.0  7304.0  6323.0  5154.0   NaN     NaN         NaN   
9   2866.0  2807.0  2806.0  2901.0  3267.0  2668.0   NaN     NaN         NaN   
10  4682.0  4869.0  4784.0  5231.0  4318.0  3072.0   NaN     NaN         NaN   
11   241.0   266.0   127.0  1156.0   341.0   129.0   NaN  bosque         NaN   
12   100.0   254.0   142.0  1004.0   456.0   236.0   NaN  bosque         NaN   
13  5969.0  6137.0  6251.0  7166.0  5648.0  4379.0   NaN     NaN         NaN   
14  6500.0  6588.0  6696.0  6850.0  5156.0  4107.0   NaN     NaN         NaN   
15  1541.0  1450.0  1409.0  2846.0  1827.0  1268.0   NaN     NaN         NaN   
16  1826.0  1623.0  1593.0  1635.0  1432.0  1197.0   NaN     NaN         NaN   
17   178.0   465.0   264.0  4109.0  1557.0   534.0   NaN  bosque         NaN   
18  4673.0  4564.0  4520.0  5229.0  3554.0  2615.0   NaN     NaN         NaN   
19  1573.0  1560.0  1508.0  1522.0  1665.0  1484.0   NaN     NaN         NaN   

    cloud  ninguno    label  
0     NaN      NaN     wofs  
1     NaN      NaN     wofs  
2   cloud      NaN    cloud  
3   cloud      NaN    cloud  
4     NaN  ninguno  ninguno  
5     NaN      NaN     wofs  
6     NaN      NaN     wofs  
7     NaN      NaN     wofs  
8   cloud      NaN    cloud  
9   cloud      NaN    cloud  
10  cloud      NaN    cloud  
11    NaN      NaN   bosque  
12    NaN      NaN   bosque  
13  cloud      NaN    cloud  
14  cloud      NaN    cloud  
15  cloud      NaN    cloud  
16  cloud      NaN    cloud  
17    NaN      NaN   bosque  
18  cloud      NaN    cloud  
19  cloud      NaN    cloud

In [11]:
# Se crea el conjunto de clasificacion y el de entrenamiento
Y = satellite["label"]
X = satellite.drop(columns=["label","wofs","bosque","wofs_bosque","cloud","ninguno"], axis=1)
Y.value_counts()

cloud          538313
wofs           288423
bosque          99236
ninguno         61378
wofs_bosque       300
Name: label, dtype: int64

# Se crean los conjuntos de entrenamiento y de prueba
Para hacer esto se deben dividir los datos en 4 conjuntos de datos de la siguiente manera: Conjuntos de entrenamiento(X_train, y_train) y Conjunto de test(X_test, y_test)

In [12]:
# Se dividen los datos en los conjuntos de Train y test
X_train, X_test, y_train, y_test = train_test_split(
    X,
    Y,
    test_size=0.3,
    random_state=0
)

In [13]:
X_test

blue    green      red      nir   swir1   swir2
441018  4415.0   4694.0   4728.0   4888.0  4886.0  4312.0
738664  3464.0   3533.0   3523.0   4645.0  3161.0  2128.0
472244   393.0    334.0    238.0    220.0   319.0   257.0
574853  5000.0   5063.0   5114.0   5233.0  5074.0  4151.0
434050  2950.0   2903.0   2836.0   3711.0  2514.0  1761.0
728385   212.0    377.0    214.0    161.0   127.0   105.0
177393   154.0    438.0    219.0   3341.0  1372.0   484.0
848812  1212.0   1279.0   1127.0   3254.0  1728.0  1051.0
489892  5699.0   5788.0   5813.0   5858.0  5049.0  3801.0
952757  7983.0   8574.0   8785.0   9480.0  6711.0  4588.0
524790  2067.0   1554.0   1671.0   1960.0  1702.0  1434.0
889190  5768.0   5886.0   5983.0   6213.0  5807.0  4848.0
871177   692.0    645.0    462.0    359.0   399.0   301.0
401322  2211.0   2258.0   2219.0   2227.0  2425.0  2168.0
17962    707.0    686.0    592.0   2030.0   975.0   566.0
226096  6664.0   6719.0   6793.0   7284.0  5152.0  3643.0
904345   321.0    223.0    113.0     90.0   121.0    95.0
250812  4021.0   3952.0   3793.0   4332.0  2480.0  1688.0
396424  1979.0   2077.0   1982.0   1936.0  2234.0  2037.0
796464  4718.0   4651.0   4786.0   5221.0  4094.0  3533.0
971028  1619.0   1539.0   1520.0   1596.0  1834.0  1513.0
396610  8274.0   8414.0   8602.0   8955.0  4323.0  2848.0
115535  6815.0   6670.0   6694.0   6611.0  2004.0  1965.0
592753  1214.0   1213.0   1083.0   2936.0  1679.0   999.0
961986  2675.0   2827.0   2722.0   4024.0  2985.0  2072.0
947227  9690.0  10220.0  10511.0  11109.0  6553.0  4384.0
593843  3753.0   3846.0   3805.0   4947.0  4030.0  3141.0
637936   234.0    249.0    139.0    132.0   169.0   140.0
819197  8429.0   8564.0   8610.0   8668.0  4626.0  3158.0
610996   126.0    152.0     39.0    813.0   250.0    83.0
...        ...      ...      ...      ...     ...     ...
879536  9211.0   9371.0   9556.0   9527.0  5176.0  3555.0
403125  3211.0   3290.0   3315.0   3452.0  3598.0  3306.0
278461   162.0    102.0     17.0      4.0    49.0    45.0
244099   240.0    148.0     53.0     37.0    77.0    62.0
745752  1726.0   1630.0   1546.0   1550.0  1714.0  1392.0
927116   120.0    391.0    172.0   4053.0  1784.0   652.0
759299   142.0    251.0    190.0    214.0   115.0    69.0
327497   347.0    226.0    124.0     95.0   115.0    90.0
700841  6570.0   6674.0   6767.0   6946.0  5307.0  3755.0
889866  6954.0   6905.0   6913.0   7034.0  2555.0  1992.0
470887  5318.0   5909.0   5794.0   5705.0  3763.0  2342.0
539041  4381.0   3851.0   3808.0   3808.0   975.0   457.0
716410  7882.0   7651.0   7749.0   7788.0  2373.0  1386.0
17805    370.0    259.0    142.0    111.0   139.0   110.0
827759  4218.0   4342.0   4302.0   5372.0  3705.0  2636.0
717575  8259.0   8618.0   8689.0   8741.0  4916.0  2955.0
440067   249.0    204.0    102.0     83.0   108.0    85.0
931976  6134.0   6378.0   6459.0   7169.0  5811.0  4301.0
266738  4422.0   4327.0   4396.0   4526.0  4274.0  3816.0
818441  1143.0    926.0    817.0    762.0   681.0   595.0
917705   369.0    284.0    161.0    125.0   149.0   122.0
935183  3573.0   3377.0   3331.0   5052.0  2899.0  2180.0
7543    4385.0   4235.0   4301.0   4403.0  4278.0  4144.0
609188   165.0    387.0    212.0   3128.0  1297.0   476.0
73642    199.0    140.0     35.0     11.0    53.0    41.0
518086   430.0    299.0    187.0    148.0   159.0   130.0
950980   112.0    293.0    162.0   1484.0   748.0   378.0
712658  4317.0   4643.0   4346.0   4060.0  2553.0  1465.0
921580  1513.0   1520.0   1432.0   1400.0  1641.0  1449.0
234064  4288.0   4478.0   4556.0   5162.0  3723.0  2976.0

[296295 rows x 6 columns]

In [14]:
y_test

441018      cloud
738664      cloud
472244       wofs
574853      cloud
434050      cloud
728385       wofs
177393     bosque
848812      cloud
489892      cloud
952757      cloud
524790      cloud
889190      cloud
871177       wofs
401322      cloud
17962     ninguno
226096      cloud
904345       wofs
250812      cloud
396424      cloud
796464      cloud
971028      cloud
396610      cloud
115535      cloud
592753      cloud
961986      cloud
947227      cloud
593843      cloud
637936       wofs
819197      cloud
610996     bosque
           ...   
879536      cloud
403125      cloud
278461       wofs
244099       wofs
745752      cloud
927116     bosque
759299       wofs
327497       wofs
700841      cloud
889866      cloud
470887      cloud
539041      cloud
716410      cloud
17805        wofs
827759      cloud
717575      cloud
440067       wofs
931976      cloud
266738      cloud
818441      cloud
917705       wofs
935183      cloud
7543        cloud
609188     bosque
73642     

# Naive Bayes para multiples capas
La clasificación multicapa del algoritmo Naive Bayes es conocida como "Clasificación Naive Bayes Multinomial". Como se puede apreciar la clasificación de nuestro conjunto de caracteristicas puede tener 4 valores: wofs, bosque, wofs_bosque y ninguno. 
En esta parte de creación del modelo se evaluaran dos variaciones del clasificador Naive Bayes: Gausiano y multinomial. En el analisis de resultados se contrastaran los resultados obtenidos para cada uno de ellos.

In [15]:
#Create a Gaussian Classifier
gnb = GaussianNB()
#gnb = MultinomialNB()

In [16]:
#Train the model using the training sets
gnb.fit(X_train, y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [17]:
#Predict the response for test dataset
y_pred = gnb.predict(X_test)

In [18]:
Counter(y_pred)

Counter({'cloud': 135465,
         'wofs': 84790,
         'bosque': 29890,
         'ninguno': 44934,
         'wofs_bosque': 1216})

In [19]:
confusion_matrix(y_test, y_pred)

array([[ 27133,     74,   1321,    349,    718],
       [    68, 134215,  27495,      1,      0],
       [  2570,   1176,  13427,   1065,    105],
       [    95,      0,   2685,  83335,    385],
       [    24,      0,      6,     40,      8]], dtype=int64)

In [20]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

      bosque       0.91      0.92      0.91     29595
       cloud       0.99      0.83      0.90    161779
     ninguno       0.30      0.73      0.42     18343
        wofs       0.98      0.96      0.97     86500
 wofs_bosque       0.01      0.10      0.01        78

    accuracy                           0.87    296295
   macro avg       0.64      0.71      0.65    296295
weighted avg       0.94      0.87      0.89    296295



In [21]:
# Model Accuracy
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.8711520612902681


In [22]:
#toBePersisted = dict({
#    'metadata': {
#        'name': 'NB_Multinomial'
#    }
#})
#dump(toBePersisted, 'NB_Multinomial.joblib')

In [23]:
# joblib.dump(gnb, 'NB_Multinomial.joblib') 
# joblib.dump(gnb, 'model_nb_multinomial.joblib') 

## Conclusiones

El rendmiento obtenido con el modelo Multinomial llega a ser del 79.73% y el rendimiento con el modelo Gausiano alcanza el 87.11%. Por lo que logramos determinar que en este escenario funciona mejor el modelo Gassiano que el Multinomial por aproimadamente 8 puntos de diferencia. Revisando la matriz de confusión logramos evidenciar que el modelo Multinomial confunde mucho de los elementos de la clase 3(Ninguno) siendo esto logico porque los valores de caracteristicas de entrada


**Multinomial**

Accuracy: 0.7973101132317454

array([[ 28721,      1,    320,      2,    551],
       [    43, 120350,  39851,    240,   1295],
       [  1895,   8902,   4897,   1003,   1646],
       [     0,   3826,    175,  82221,    278],
       [     0,     12,      7,      9,     50]], dtype=int64)
       
**Gaussiano**

Accuracy: 0.8711520612902681

array([[ 27133,     74,   1321,    349,    718],
       [    68, 134215,  27495,      1,      0],
       [  2570,   1176,  13427,   1065,    105],
       [    95,      0,   2685,  83335,    385],
       [    24,      0,      6,     40,      8]], dtype=int64)

# Referencias
https://www.datacamp.com/community/tutorials/naive-bayes-scikit-learn

https://datascience.stackexchange.com/questions/40345/how-to-convert-multiple-columns-into-single-columns-in-pandas

https://stackoverflow.com/questions/13295735/how-can-i-replace-all-the-nan-values-with-zeros-in-a-column-of-a-pandas-datafra